rust bindings to kvmi apis

arnabcs17b006 opened the pull request.

arnabcs17b006 committed:
Update lib.rs

Commit:
provision to read msrs registers

Commit:
read msrs

Commit:
minor fixes

Commit:
write operation on registers

write operation on registers

minor change

revert back

minor fix

Commit:
adding write_physical

Commit:
msr events

msr events

msr events

minor fix

minor fix

minor fix

corrected msr indices

formatting

mock added

upgrading mocktopus version

PartialEq added to KVMiCr and KVMiMsr

changed mocking framework to mockall

replace mocktopus with mockall as it runs on rust stable

removed mockall

singlestep added

removed single step as it is not supported in kvmi-v6

Commit:
interrupt event

Commit:
gpa and insn_len for breakpoint added

Commit:
Page fault event added

Pagefault event added

interrupt events added

Commit:
memory vevent reply added

memory event reply added

x86 crate added

formatting done

removed x86 as it requires nightly

suppresing clippy errors

clippy errors resolved

rustfmt used

conflict resolved

minor fix

correcting silly mistake

Wenzel commented on Cargo.toml.save:
```diff
@@ -0,0 +1,21 @@
+[package]
```
What is this file ?

Wenzel commented on src/lib.rs:
```diff
@@ -197,20 +268,38 @@ impl KVMi {
+        let mut msrs: kvm_msr = unsafe { mem::MaybeUninit::<kvm_msr>::zeroed().assume_init() };
         let mut mode: c_uint = 0;
+        msrs.msrs.nmsrs=6;
+        msrs.entries[0].index = 0x00000174;
```
Let's not use hardcoded indexes here.
Please use the x86 crate that I mentionned in Slack

arnabcs17b006 replied:
It seems x86 crate works only with nightly

Wenzel reviewed (changes requested):
@arnabcs17b006 please see my comments.

Also please improve the readability of your commits:
- `minor fix` doesn't mean anything
- `revert back`
- `minor change`

etc.

you can squash them if you want.

Wenzel commented:
@arnabcs17b006 could you rebase on `master` and repush this branch ?
we integreated a CI workflow for this crate, to check if everything compiles.

arnabcs17b006 commented:
It seems x86 crate works only with nightly

Wenzel commented on Cargo.toml:
```diff
@@ -17,3 +18,9 @@ num-traits = "0.2.8"
 libc = "0.2.66"
 libloading = "0.6.1"
+mockall = "0.7.1"
+clippy = {version = "0.0.188", optional = true}
```
Do we need to add clippy as a dependency ?
it's not a strong dependency of kvmi, only a development tool, used by the CI.
opinion @rageagainsthepc  ?

rageagainsthepc replied:
If it's not used in the code it shouldn't be a dependency.

Wenzel commented on Cargo.toml:
```diff
@@ -17,3 +18,9 @@ num-traits = "0.2.8"
+clippy = {version = "0.0.188", optional = true}
+
+[features]
+default = []
```
Do we need a empty features section ?

Wenzel commented on src/lib.rs:
```diff
@@ -2,36 +2,29 @@
-use std::sync::{Condvar, Mutex};
-
 use enum_primitive_derive::Primitive;
+use kvmi_sys::kvm_msr_entry;
```
you can group this import with the `use kvmi_sys::` block below

Wenzel commented on src/lib.rs:
```diff
@@ -58,20 +78,36 @@ pub enum KVMiEventReply {
+#[derive(Primitive, Debug, Copy, Clone, PartialEq)]
 pub enum KVMiCr {
     Cr0 = 0,
+    Cr2 = 2,
```
Again, remove CR2 here, it is not interceptable.

Wenzel commented on src/lib.rs:
```diff
@@ -172,6 +215,14 @@ impl KVMi {
         Ok(())
     }
 
+    pub fn control_msr(&self, vcpu: u16, reg: KVMiMsr, enabled: bool) -> Result<(), Error> {
```
I understand the idea of wrapping the MSR with an enum, but by doing so we narrow the possibilities of this  API.
In fact, the user should be able to pass any MSR index, not only those exposed by the KVMiMsr struct.
so `reg` should be an `u32` here.
cc @rageagainsthepc

rageagainsthepc replied:
That may be my fault because I remember suggesting an enum instead of an integer some weeks ago. I don't know though if it had something to do with msrs. @Wenzel You are right that an enum does not make sense for msrs because of their fluctuating nature. @arnabcs17b006 Sorry about that.

Wenzel reviewed (changes requested):
Please see my comments.

Commit:
worked on PR reviews

worked on PR reviews

new type KVMiRegs introduced

changed type from kvm_regs to KVMiRegs in set_registers()

Wenzel commented on src/lib.rs:
```diff
@@ -197,20 +277,40 @@ impl KVMi {
         Ok((regs, sregs, msrs))
     }
 
+    pub fn set_registers(&self, vcpu: u16, regs: &kvm_regs) -> Result<(), Error> {
```
Actually, the problem that we face when implementing https://github.com/Wenzel/libmicrovmi/pull/79/files#diff-ed6d1a9ce6a167171bc40a915157b89fR140 is because we lack some abstractions from the kvmi crate.

I missed this in my review, but this public API is exposing the type `kvm_regs` from kvmi_sys.
As you defined a `pub type KVMiRegs = kvm_regs;` for this, you should reuse this type definition in this function signature too to be consistent.

Wenzel commented:
Thanks !

rageagainsthepc commented on Cargo.toml:
```diff
@@ -17,3 +18,6 @@ num-traits = "0.2.8"
 nix = "0.16.1"
 libc = "0.2.66"
 libloading = "0.6.1"
+mockall = "0.7.1"
```
This crate is not used.
